Allow users to click account renewal links multiple times without hitting an 'Invalid Token' page

anoadragon453 opened the pull request:
This PR makes some modifications to the account renewal endpoint to enable idempotency. For context, when the [account validity](https://github.com/matrix-org/synapse-dinsic/blob/854764073c984da8d7f5daf713a550b2c9691f11/docs/sample_config.yaml#L1195-L1208) feature is in use, users are sent an email with a renewal link a little while before their account expires.

Users may end up clicking this link twice accidentally, or their spam software will click it for them before they have a chance to do so manually. In each case, every click after the first would return an Invalid Token error page, as we delete renewal tokens once used.

This PR changes things up a bit. We no longer delete the token upon use. However, we also ensure that tokens cannot be re-used by ~~(ab)using the `email_sent` boolean. This boolean is set to True when an email with the token is sent out, and set to False when the account is renewed using that token. With this, and with the token still intact,~~ adding a new `token_used_ts_ms` column to the `account_validity` table. With this, we can check if the token is valid, yet has already been used. In this case, return a new template which simply informs the user of when their account will expire, but not that it was just renewed.

Part of this PR was also adding information about when a user's account would expire in the actual renewal template. Mostly for nicer UX purposes. Doing so meant converting the account validity config code to using the new `read_templates` function, which simplifies it massively!

Reviewable commit-by-commit.

anoadragon453 committed:
Don't clear a user's renewal token after they've validated their account

Also some typing and docstring fixups.

anoadragon453 committed:
Switch account_validity templates to read_template method

We're doing this as we're planning to inject account expiry dates into the templates
(so that they can show "Your account is valid until XX/YY/ZZ"). Since this is
only information we'll have during server operation, we need to switch the config
variables to be templates instead.

Also nice as we no longer need to hardcoded the template content inline.

The public_baseurl change was due to self.public_baseurl raising an
AssertionError if we try to call it before it exists, which happens to
be the case for the account validity config code. Note that we override
__getattr__ above.

anoadragon453 committed:
Add template for trying to renew an account with a previously used token

For when users accidentally click the renewal link twice.

anoadragon453 committed:
Return the new template if the user is trying to reuse a token.

We check if the user is re-using a token by (ab)using the 'email_sent' column associated with
each user's account validity information. When a token is first created and emailed to a user,
'email_sent' is set to True. When a user comes back to us with this token, we renew the account
and set 'email_sent' to False - but critically we keep the token around.

If the user comes back again with the same token, we'll see that although the token is valid,
'email_sent' is False. From this we can learn that the token was already used to renew the
account. We call this case a stale token, and return a new template to the user. The new
template simply informs the user when their account will expire.

This is much nicer UX than the 'invalid token!' users would get when they tried a token
twice.

anoadragon453 committed:
Modify reg tests to check re-using an existing account validity token.

Note that the variable for this test include an account expiring after 1 week of it
last being renewed (or created), and sending the user a renewal token 2 days before
the account is set to expire.

The change here is that after we successfully renew the account, move forward and
try to renew it again. We should get a success, and the new template returned,
but the account's expiration date should not have been extended.

clokep commented:
> In this case, return a new template which simply informs the user of when their account with expire, but not that it was just renewed.

Curious why this was done? Why not just return the same template always?

giomfo commented:
@clokep: the plan is to be able to handle separately the 2 cases: 
- the request renews actually the user's account
- the request do nothing (because the provided token has already been used)

clokep commented:
Right, but it seems like a generic message could be used in both cases "Your account renewal succeeded and is valid until xxx yy, zzzz."

I haven't thought through all the scenarios, but I'm wondering if there's an attack by showing different information (similar to how you limit the information during password resets). I doubt there is, but wanted to mention it!

giomfo commented:
I understand what you mean, the aim here is to keep the opportunity to customize the 2 cases. 
The second case may happen at any time (whereas the account has been already renewed several days ago), we want to be sure the user understands correctly the message.

@anoadragon453 I would say if there is no technical problem to handle separately these 2 cases (And I think this is already done), I would prefer handle them separately. If you see any reason to merge these 2 cases, feel free to merge them

clokep commented on synapse/res/templates/account_renewed.html:
```diff
@@ -1 +1 @@
-<html><body>Your account has been successfully renewed.</body><html>
+<html><body>Your account has been successfully renewed until {{ expiration_ts|format_ts("%d-%m-%Y") }}.</body><html>
```
```suggestion
<html><body>Your account has been successfully renewed and is valid until {{ expiration_ts|format_ts("%d-%m-%Y") }}.</body><html>
```

clokep commented on synapse/storage/databases/main/registration.py:
```diff
@@ -224,22 +224,29 @@ async def set_renewal_token_for_user(
         Returns:
-            The ID of the user to which the token belongs.
+            A tuple of containing the following values:
+                * The ID of a user to which the token belongs.
+                * Whether an email has been sent to the user with the token, and
+                    that they haven't renewed their account with it yet.
```
```suggestion
                * The ID of a user to which the token belongs.
                * Whether an email has been sent to the user with the token, and
                    that they haven't renewed their account with it yet.
                * An int representing the user's expiry timestamp as milliseconds since the
                    epoch, or 0 if the token was invalid.
```

clokep commented on synapse/handlers/account_validity.py:
```diff
@@ -223,37 +223,70 @@ async def _get_renewal_token(self, user_id: str) -> str:
+        # True if a user is found and email_sent is False, meaning the token has already
+        # been used. False otherwise.
+        if not email_sent:
+            logger.warning(
+                "User %s attempted to use previously used token to renew account",
+                user_id,
+            )
```
Is this really a warning? Are we considering this potentially malicious?

It is probably worth including the token in the log line?

anoadragon453 replied:
I considered it useful to log in case a user asks why their renewal date isn't updating. But this information could be gleamed from checking what template they were receiving + looking at request logs and their `account_validity` table entry.

If you're asking about warning versus info, yeah perhaps a *warning* doesn't quite make sense here.

clokep replied:
I was more asking about warning vs. info. Logging something is reasonable. 😄

anoadragon453 replied:
Gotcha. Changed to `info` and am now logging the token.

clokep commented on synapse/handlers/account_validity.py:
```diff
@@ -223,37 +223,70 @@ async def _get_renewal_token(self, user_id: str) -> str:
-        await self.renew_account_for_user(user_id)
 
-        return True
+        # Renew the account. Keep the renewal token in tact for idempotency in case the
+        # user attempts to renew their account with the same token.
```
```suggestion
        # Renew the account. Keep the renewal token intact for idempotency in case the
        # user attempts to renew their account with the same token.
```

clokep commented on tests/rest/client/v2_alpha/test_register.py:
```diff
@@ -706,7 +706,32 @@ def test_renewal_email(self):
+
+        # Check that we're getting HTML back.
+        content_type = None
+        for header in channel.result.get("headers", []):
+            if header[0] == b"Content-Type":
+                content_type = header[1]
```
`channel` should have a `headers` property on it which returns a `Headers` instance, I think using that would simplify this code a bit.

clokep commented on synapse/config/_base.py:
```diff
@@ -244,7 +244,7 @@ def read_templates(
         # Update the environment with our custom filters
         env.filters.update({"format_ts": _format_ts_filter})
-        if self.public_baseurl:
+        if self.root and self.root.public_baseurl:
```
I don't understand why this change is necessary.

anoadragon453 replied:
Oh, by:

> The public_baseurl change was due to self.public_baseurl raising an
AssertionError if we try to call it before it exists, which happens to
be the case for the account validity config code. Note that we override
__getattr__ above.

in the commit message I was referring to this bit, but that probably wasn't too clear.

Essentially when trying to call `read_templates` from `AccountValidityConfig.__init__`, I found that `public_baseurl` had not been set yet, and thus this check would fail. This fix got around the issue, and worked since the account validity code didn't include any MXC URLs. But... of course it could in the future.

I just took another look and realised that the issue is due to calling `AccountValidityConfig.read_templates` instead of `RegistrationConfig.read_templates`. In the former, `public_baseurl` is not set, but in the latter it is. I've pushed 58300ee9b to fix this with a little more explanation. Let me know what you think about the solution.

clokep reviewed (commented):
The code seems like it would work ok.

> We no longer delete the token upon use. However, we also ensure that tokens cannot be re-used by (ab)using the email_sent boolean. This boolean is set to True when an email with the token is sent out, and set to False when the account is renewed using that token.

I'm not a big fan of this approach -- multiple pieces of state ending up in a boolean isn't great. Can we add an additional boolean column of "token_used" or something? (Or maybe a timestamp of when it was first clicked.)

anoadragon453 committed:
Move account validity template reading to RegistrationConfig

Calling `self.read_templates` from AccountValidityConfig does not work as
expected as it does not have state such as self.public_baseurl properly set up.
This is due to AccountValidityConfig not being initialised like other config
classes, but instead simply in the body of RegistrationConfig.read_config.

I think there's an argument for initialising it properly, instead of as
a part of registration config. Perhaps even putting it in its own file.
But that should be left to a separate PR.

anoadragon453 committed:
Apply suggestions from code review

Co-authored-by: Patrick Cloke <clokep@users.noreply.github.com>

anoadragon453 committed:
Change log line to be info, also log attempted token

anoadragon453 committed:
Simplify Content-Type checking test code

anoadragon453 committed:
Add token_used_ts_ms column to account_validity table

anoadragon453 committed:
Switch to a new token_used_ts_ms db column, instead of overloading email_sent

anoadragon453 committed:
mypy

anoadragon453 commented:
> Can we add an additional boolean column of "token_used" or something? (Or maybe a timestamp of when it was first clicked.)

The original reason for using `email_sent` was to minimise the necessary changes for this feature. But given the current size of this PR I don't really have any objections with adding a migration.

I've added a column `token_used_ts_ms` to `account_validity`, which we now use to check the stale status of a token.

anoadragon453 committed:
Merge branch 'dinsic' of github.com:matrix-org/synapse-dinsic into anoa/account_renewal_idempotency

clokep commented on synapse/handlers/account_validity.py:
```diff
@@ -223,47 +223,84 @@ async def _get_renewal_token(self, user_id: str) -> str:
-        await self.renew_account_for_user(user_id)
 
-        return True
+        # Renew the account. Keep the renewal token intact for idempotency in case the
+        # user attempts to renew their account with the same token.
```
I find this comment a bit confusing -- what does "keeping it intact" mean? Why would we not be doing that anyway?

anoadragon453 replied:
After a re-read I agree with you.

By "keeping it intact", I mean "do not clear the token on account renewal". I'll reword this.

clokep commented on synapse/storage/databases/main/registration.py:
```diff
@@ -220,26 +224,43 @@ async def set_renewal_token_for_user(
-            The ID of the user to which the token belongs.
+            A tuple of containing the following values:
+                * The ID of a user to which the token belongs.
+                * Whether an email has been sent to the user with the token, and
+                    that they haven't renewed their account with it yet.
```
Why would an email not have been sent? Is this return value just whether they've renewed or not?

anoadragon453 replied:
Err, this explanation was part of using the `email_sent` boolean to determine whether the current token had been used to renew the account. With switching to `token_used_ts_ms` this text can be simplified.

This bool would be false if the user had renewed their account, and an email has not been sent since then. Otherwise it is set to true upon an email being sent out.

In the grand scheme of things we probably could have just renamed `email_sent` to `token_unused`, but that doesn't feel as clean as having a timestamp of when the user renewed their account with the given token.

anoadragon453 replied:
Oh, we're not even using `email_sent` anymore huh. I'll remove this from the return tuple.

clokep commented on synapse/config/registration.py:
```diff
@@ -98,6 +73,30 @@ def read_config(self, config, **kwargs):
         )
 
+        # Load account validity templates.
+        # We do this here instead of in AccountValidityConfig as read_templates
+        # relies on state that hasn't been initialised in AccountValidityConfig
```
This is due to the `AccountValidityConfig` not being part of the huge config object that normally gets created? I wonder if there's a good reason for that.

anoadragon453 replied:
Indeed, and my thought is that this class was just created to clean up `RegistrationConfig` a bit, rather than making `AccountValidityConfig` a proper config class.

But I'm not sure if there's any justification against not doing so. May have to wait for people to come back to ask.

clokep replied:
We talked a bit about this and the conclusion was to attempt to make it a proper config class, but if it turns out to be too hard then to punt it.

clokep reviewed (commented):
Overall this seems OK. I think I want to discuss the config bits a bit more though.

anoadragon453 committed:
Remove email_sent from get_user_from_renewal_token return values

We're not using this to calculate whether a user has used this token previously
anymore.

clokep commented on synapse/storage/databases/main/registration.py:
```diff
@@ -256,7 +254,6 @@ async def get_user_from_renewal_token(
 
         return (
             ret_dict["user_id"],
-            ret_dict["email_sent"],
```
We should not request this column from the database above. 👍

anoadragon453 committed:
Remove unnecessary email_sent column from query

anoadragon453 committed:
Move AccountValidityConfig to its own config file

anoadragon453 committed:
Change account validity config option instances

anoadragon453 commented on synapse/config/account_validity.py:
```diff
@@ -0,0 +1,146 @@
+from synapse.config._base import Config, ConfigError
+
+
+class AccountValidityConfig(Config):
```
This diff doesn't make it obvious what changed in the move. Essentially it was just a cut n' paste, but setting up `read_config` (instead of `__init__`) and `generate_config_section` methods for `AccountValidityConfig`.

I also modified `self.account_validity.xxx` instances to `self.account_validity_xxx`. Stuff that pulled from `config` (which used to be `synapse_config["account_validity"]`, now pulls from new var `account_validity_config`.

anoadragon453 commented on synapse/config/account_validity.py:
```diff
@@ -0,0 +1,146 @@
+        # after that the validity period changes and Synapse is restarted, the users'
+        # expiration dates won't be updated unless their account is manually renewed. This
+        # date will be randomly selected within a range [now + period - d ; now + period],
+        # where d is equal to 10% of the validity period.
```
I needed to switch instances of `%%` to `%` in this config block. This is because we were running `% locals()` on this text before in `RegistrationConfig`, but we do not need to here. All of the placeholder variables were related to registration options.

The `%(app)s` bits are meant to be left in the config, and will get parsed by other code during runtime.

anoadragon453 committed:
Correct test.

Since we're pulling config options out of the config during class
__init__ functions now, instead of during runtime,
we can't just change the homeserver config and expect runtime to
have the new value.

We need to change the vars that get set in those __init__ functions
directly. In this case, RegistrationStore.__init__.

anoadragon453 reviewed (commented):
Updated to make `AccountValidityConfig` its own actual config class.

Let me know if you want this to be in a separate PR instead :slightly_smiling_face:

clokep commented on tests/rest/client/v2_alpha/test_register.py:
```diff
@@ -863,7 +863,7 @@ def make_homeserver(self, reactor, clock):
 
         self.hs = self.setup_test_homeserver(config=config)
-        self.hs.config.account_validity_period = self.validity_period
+        self.hs.get_datastore()._account_validity_period = self.validity_period
```
Would a better way to do this be to define `default_config`?

anoadragon453 replied:
`hs.config.account_validity_period` still won't be set even if `account_validity.period` if defined in the homeserver config though, right?

anoadragon453 replied:
We talked a bit in #synapse-dev about this. The reason this test was working before is because we were reading `period` and `startup_job_max_delta` from the homeserver config file during test execution.

Now we're reading account validity config values during homeserver initialisation, which will happen during the `self.setup_test_homeserver` call. *Before* setting those homeserver config options.

I've added a comment to explain this, but this test probably still needs a separate look. (For instance, why does the docstring of `test_background_job` say it does the same thing as `test_background_job`??).

clokep commented on synapse/handlers/account_validity.py:
```diff
@@ -37,29 +37,38 @@ def __init__(self, hs):
         self.profile_handler = self.hs.get_profile_handler()
 
+        if self._account_validity_enabled:
+            self._account_validity_period = self.hs.config.account_validity_period
```
It might be clearer to set `account_validity_period` to `None` if account validity is not enabled in the config file, but up to you if you want to make the change.

anoadragon453 replied:
Yeah, there's actually a few instances of this, and doing so would probably be clearer.

clokep commented on synapse/handlers/account_validity.py:
```diff
@@ -223,47 +232,87 @@ async def _get_renewal_token(self, user_id: str) -> str:
         Returns:
-            Whether the provided token is valid.
+            A tuple containing:
+              * A bool representing whether the token is valid.
```
```suggestion
              * A bool representing whether the token is valid and unused.
```

clokep reviewed (approved):
I think this looks fine.

anoadragon453 committed:
Update synapse/handlers/account_validity.py

Co-authored-by: Patrick Cloke <clokep@users.noreply.github.com>

anoadragon453 committed:
Allow account_validity to be None in the config

anoadragon453 committed:
Define AV variables as None initially, instead of not setting them

Also localise a variable that didn't need to be class-level.

anoadragon453 committed:
Add some explanation to directly setting account_validity-related config

anoadragon453 commented:
This PR was never ported to mainline unfortunately. After a bit of cleanup, a port has been put up: https://github.com/matrix-org/synapse/pull/9832
